# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/Zone.h
# Commit: b6d68da027a2
# Full Hash: b6d68da027a23da0a86fd73efc2fc004a2e763ba
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2023-07-04 21:49:05
# Regressor Bug: 1837620
# File Overlap Count: 1
# Description:
#   Bug 1837620 - Part 6: Make edges for multiple shape guard weak too r=sfink
#   
#   To make the edges for the multiple shape guard weak we need a way to find and
#   update the objects used to hold these edges. The simplest way to do this is to
#   add a vector in the zone.  I don't believe that there are not huge numbers of
# ==============================================================================

diff -r 318af9553464 -r b6d68da027a2 js/src/gc/Zone.h
--- a/js/src/gc/Zone.h	Tue Jul 04 13:58:57 2023 +0000
+++ b/js/src/gc/Zone.h	Tue Jul 04 13:58:57 2023 +0000
@@ -309,6 +309,13 @@
   friend class js::WeakRefObject;
   js::MainThreadOrGCTaskData<KeptAliveSet> keptObjects;
 
+  // To support weak pointers in some special cases we keep a list of objects
+  // that need to be traced weakly on GC. This is currently only used for the
+  // JIT's ShapeListObject. It's assumed that there will not be many of these
+  // objects.
+  using ObjectVector = js::GCVector<JSObject*, 0, js::SystemAllocPolicy>;
+  js::MainThreadOrGCTaskData<ObjectVector> objectsWithWeakPointers;
+
  public:
   static JS::Zone* from(ZoneAllocator* zoneAlloc) {
     return static_cast<Zone*>(zoneAlloc);
@@ -344,6 +351,9 @@
 
   void traceWeakJitScripts(JSTracer* trc);
 
+  bool registerObjectWithWeakPointers(JSObject* obj);
+  void sweepObjectsWithWeakPointers(JSTracer* trc);
+
   void addSizeOfIncludingThis(mozilla::MallocSizeOf mallocSizeOf,
                               JS::CodeSizes* code, size_t* regexpZone,
                               size_t* jitZone, size_t* baselineStubsOptimized,